pkg/skaffold/walk/walk.go

```diff
@@ -0,0 +1,165 @@
+/*
+Copyright 2019 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package walk
+
+import (
+	"os"
+
+	"github.com/karrick/godirwalk"
+)
+
+type Dirent interface {
+	IsDir() bool
+	Name() string
+}
+
+type Predicate func(string, Dirent) bool
+
+type Action func(string, Dirent) error
+
+type Excluder interface {
+	Exclude(relPath string) (bool, error)
+	SkipDir(relPath string) bool
+}
+
+type Builder interface {
+	// Options
+	Unsorted() Builder
+
+	// Predicates
+	When(Predicate) Builder
+	WhenIsDir() Builder
+	WhenIsFile() Builder
+	WhenHasName(string) Builder
+
+	// Actions
+	Do(Action) error
+	MustDo(Action)
+	AppendPaths(*[]string) error
+	CollectPaths() ([]string, error)
+}
+
+type builder struct {
+	dir       string
+	unsorted  bool
+	predicate Predicate
+}
+
+func From(dir string) Builder {
+	return &builder{
+		dir:       dir,
+		unsorted:  false,
+		predicate: func(string, Dirent) bool { return true },
+	}
+}
+
+func (w *builder) Unsorted() Builder {
+	w.unsorted = true
+	return w
+}
+
+func (w *builder) When(predicate Predicate) Builder {
+	w.predicate = and(w.predicate, predicate)
+	return w
+}
+
+func (w *builder) WhenIsFile() Builder {
+	return w.When(isFile)
+}
+
+func (w *builder) WhenIsDir() Builder {
+	return w.When(isDir)
+}
+
+func (w *builder) WhenHasName(name string) Builder {
+	return w.When(hasName(name))
+}
+
+func (w *builder) AppendPaths(values *[]string) error {
+	return w.Do(appendPaths(values))
+}
+
+func (w *builder) CollectPaths() ([]string, error) {
+	var paths []string
+	err := w.Do(appendPaths(&paths))
+	return paths, err
+}
+
+func (w *builder) Do(action Action) error {
+	info, err := os.Lstat(w.dir)
+	if err != nil {
+		return err
+	}
+
+	if !info.IsDir() {
+		if w.predicate(w.dir, info) {
+			if err := action(w.dir, info); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+
+	return godirwalk.Walk(w.dir, &godirwalk.Options{
+		Unsorted: w.unsorted,
+		Callback: func(path string, info *godirwalk.Dirent) error {
+			if w.predicate(path, info) {
+				if err := action(path, info); err != nil {
+					return err
+				}
+			}
+			return nil
+		},
+	})
+}
+
+func (w *builder) MustDo(action Action) {
+	if err := w.Do(action); err != nil {
+		panic("unable to list files")
```

Comment:
I don't feel strongly about this but go/go-read#dont-panic suggests using `log.Fatal()` rather than panicking, but I can see this would be useful.  Regardless, we should include the error here.

Reply:
`panic` is how it's done in the stdlib. See `regexp.MustCompile` for example. I'll add the error.